tests/integration/test_pip_packages.py

```diff
@@ -24,3 +28,184 @@ def test_failing_pip_local_path(test_env):
     assert completed_response.data["state"] == "failed"
     error_msg = "Direct references with 'file' scheme are not supported"
     assert error_msg in completed_response.data["state_reason"]
+
+
+def test_pip_with_cached_deps(test_env, tmpdir):
```

Comment:
Can we extract all the embedded functions from this one unless they are needed to be nested here?